File: src/utils/cache.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import { join } from "path";
import { getAppConfigDir } from "./config.ts";
import {
  normalizeCourseSection,
  type MoodleCourse,
  type MoodleCourseSection,
  type MoodleUpcomingAssignment,
} from "./moodle.ts";

interface CoursesCacheData {
  timestamp: number;
  courses: MoodleCourse[];
}

interface CacheEntry<TData> {
  timestamp: number;
  data: TData;
}

interface DashboardCacheData {
  courses: MoodleCourse[];
  upcomingAssignments: MoodleUpcomingAssignment[];
}

interface CacheData {
  // Legacy layout
  timestamp?: number;
  courses?: unknown;
  // Current layout
  dashboard?: CacheEntry<unknown>;
  coursePages?: Record<string, CacheEntry<unknown>>;
}

const CACHE_TTL_MS = 1000 * 60 * 60 * 24 * 21;
const MAX_CACHED_COURSE_PAGES = 48;

function getCacheFilePath(): string {
  return join(getAppConfigDir(), "cache.json");
}

function ensureCacheDir(): void {
  mkdirSync(getAppConfigDir(), { recursive: true });
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  return value as Record<string, unknown>;
}

function asNumber(value: unknown): number | undefined {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  return undefined;
}

function asString(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  return value;
}

function isExpired(timestamp: number): boolean {
  return Date.now() - timestamp > CACHE_TTL_MS;
}

function normalizeCourses(value: unknown): MoodleCourse[] {
  if (!Array.isArray(value)) return [];

  const courses: MoodleCourse[] = [];

  for (const entry of value) {
    if (!entry || typeof entry !== "object") continue;
    const row = entry as Partial<MoodleCourse>;
    if (typeof row.id !== "number") continue;
    if (typeof row.shortname !== "string") continue;
    if (typeof row.fullname !== "string") continue;

    const course: MoodleCourse = {
      id: row.id,
      shortname: row.shortname,
      fullname: row.fullname,
    };

    if (typeof row.displayname === "string") course.displayname = row.displayname;
    if (typeof row.categoryid === "number") course.categoryid = row.categoryid;
    if (typeof row.categoryname === "string") course.categoryname = row.categoryname;
    if (typeof row.summary === "string") course.summary = row.summary;
    if (typeof row.visible === "number") course.visible = row.visible;
    if (typeof row.progress === "number" || row.progress === null) {
      course.progress = row.progress;
    }
    if (typeof row.courseurl === "string") course.courseurl = row.courseurl;

    courses.push(course);
  }

  return courses;
}

function normalizeUpcomingAssignments(value: unknown): MoodleUpcomingAssignment[] {
  if (!Array.isArray(value)) return [];

  const assignments: MoodleUpcomingAssignment[] = [];

  for (const entry of value) {
    const row = asRecord(entry);
    if (!row) continue;

    const id = asNumber(row.id);
    const dueDate = asNumber(row.dueDate);
    const courseId = asNumber(row.courseId);
    const name = asString(row.name);
    if (id === undefined || dueDate === undefined || courseId === undefined || !name) continue;

    const assignment: MoodleUpcomingAssignment = {
      id,
      name,
      dueDate,
      courseId,
    };

    const courseShortName = asString(row.courseShortName);
    const courseFullName = asString(row.courseFullName);
    if (courseShortName) assignment.courseShortName = courseShortName;
    if (courseFullName) assignment.courseFullName = courseFullName;

    assignments.push(assignment);
  }

  return assignments;
}

function normalizeCourseSections(value: unknown): MoodleCourseSection[] {
  if (!Array.isArray(value)) return [];

  const sections: MoodleCourseSection[] = [];
  for (const entry of value) {
    const normalized = normalizeCourseSection(entry);
    if (!normalized) continue;
    sections.push(normalized);
  }

  return sections;
}

function readCache(): CacheData {
  try {
    const file = getCacheFilePath();
    if (!existsSync(file)) return {};
    const raw = readFileSync(file, "utf-8");
    const parsed = JSON.parse(raw) as unknown;
    const record = asRecord(parsed);
    if (!record) return {};
    return record as CacheData;
  } catch {
    return {};
  }
}

function writeCache(cache: CacheData): void {
  ensureCacheDir();
  writeFileSync(getCacheFilePath(), JSON.stringify(cache, null, 2), { mode: 0o600 });
}

function pruneCoursePages(
  pages: Record<string, CacheEntry<unknown>> | undefined,
): Record<string, CacheEntry<unknown>> {
  if (!pages) return {};

  const entries = Object.entries(pages)
    .filter(([, entry]) => !isExpired(entry.timestamp))
    .sort((left, right) => right[1].timestamp - left[1].timestamp)
    .slice(0, MAX_CACHED_COURSE_PAGES);

  return Object.fromEntries(entries);
}

function getDashboardFromCacheValue(
  rawDashboard: CacheEntry<unknown> | undefined,
): DashboardCacheData | null {
  if (!rawDashboard) return null;
  if (isExpired(rawDashboard.timestamp)) return null;

  const dashboardRecord = asRecord(rawDashboard.data);
  if (!dashboardRecord) return null;

  const courses = normalizeCourses(dashboardRecord.courses);
  const upcomingAssignments = normalizeUpcomingAssignments(
    dashboardRecord.upcomingAssignments,
  );

  return { courses, upcomingAssignments };
}

function getLegacyCoursesFromCache(cache: CacheData): MoodleCourse[] | null {
  const legacyTimestamp = asNumber(cache.timestamp);
  if (legacyTimestamp !== undefined && isExpired(legacyTimestamp)) {
    return null;
  }

  const courses = normalizeCourses(cache.courses);
  if (courses.length > 0) return courses;
  if (Array.isArray(cache.courses)) return [];
  return null;
}

function saveCache(cache: CacheData): void {
  const normalizedCache: CacheData = {
    ...cache,
    coursePages: pruneCoursePages(cache.coursePages),
  };
  writeCache(normalizedCache);
}

export function getCachedDashboard():
  | {
      courses: MoodleCourse[];
      upcomingAssignments: MoodleUpcomingAssignment[];
    }
  | null {
  try {
    const cache = readCache();
    const dashboardEntry = getDashboardFromCacheValue(cache.dashboard);
    if (dashboardEntry) {
      return dashboardEntry;
    }

    const legacyCourses = getLegacyCoursesFromCache(cache);
    if (!legacyCourses) return null;
    return { courses: legacyCourses, upcomingAssignments: [] };
  } catch {
    return null;
  }
}

export function getCachedCourses(): MoodleCourse[] | null {
  try {
    const dashboard = getCachedDashboard();
    return dashboard ? dashboard.courses : null;
  } catch {
    return null;
  }
}

export function saveDashboardToCache(
  courses: MoodleCourse[],
  upcomingAssignments: MoodleUpcomingAssignment[],
): void {
  try {
    const cache = readCache();
    const now = Date.now();
    cache.timestamp = now;
    cache.courses = courses;
    cache.dashboard = {
      timestamp: now,
      data: {
        courses,
        upcomingAssignments,
      },
    };

    saveCache(cache);
  } catch {
    // ignore cache write failures
  }
}

export function saveCoursesToCache(courses: MoodleCourse[]): void {
  try {
    const cache = readCache();
    const cachedDashboard = getDashboardFromCacheValue(cache.dashboard);
    const upcomingAssignments = cachedDashboard?.upcomingAssignments ?? [];
    saveDashboardToCache(courses, upcomingAssignments);
  } catch {
    // ignore cache write failures
  }
}

export function getCachedCourseSections(courseId: number): MoodleCourseSection[] | null {
  try {
    const cache = readCache();
    const key = String(courseId);
    const entry = cache.coursePages?.[key];
    if (!entry) return null;
    if (isExpired(entry.timestamp)) {
      if (cache.coursePages) {
        delete cache.coursePages[key];
        saveCache(cache);
      }
      return null;
    }

    const sections = normalizeCourseSections(entry.data);
    if (sections.length > 0) return sections;
    if (Array.isArray(entry.data)) return [];
    return null;
  } catch {
    return null;
  }
}

export function saveCourseSectionsToCache(
  courseId: number,
  sections: MoodleCourseSection[],
): void {
  try {
    const cache = readCache();
    const key = String(courseId);

    cache.coursePages = {
      ...(cache.coursePages ?? {}),
      [key]: {
        timestamp: Date.now(),
        data: sections,
      },
    };

    saveCache(cache);
  } catch {
    // ignore cache write failures
  }
}

export function clearCache(): void {
  try {
    const now = Date.now();
    const payload: CacheData & CoursesCacheData = {
      timestamp: now,
      courses: [],
      dashboard: {
        timestamp: now,
        data: {
          courses: [],
          upcomingAssignments: [],
        },
      },
      coursePages: {},
    };
    writeCache(payload);
  } catch {
    // ignore
  }
}

File: src/storage/cache.rs
use crate::models::{Course, CourseSection, UpcomingAssignment};
use crate::storage::{StorageError, config_dir};
use serde_json::{Value, json};
use std::path::PathBuf;
use tui_components::storage::json::{named_file, read_json_or_default, write_json_pretty};
use tui_components::storage::time::{is_expired as timestamp_is_expired, now_ms};

const CACHE_TTL_MS: u64 = 1000 * 60 * 60 * 24 * 21;
const MAX_CACHED_COURSE_PAGES: usize = 48;

#[derive(Debug, Clone, Default)]
pub struct DashboardCacheData {
    pub courses: Vec<Course>,
    pub upcoming_assignments: Vec<UpcomingAssignment>,
}

pub fn cache_file() -> Result<PathBuf, StorageError> {
    Ok(named_file(config_dir()?, "cache.json"))
}

fn is_expired(timestamp: u64) -> bool {
    timestamp_is_expired(timestamp, CACHE_TTL_MS)
}

fn read_cache_value() -> Value {
    cache_file()
        .map(read_json_or_default)
        .unwrap_or_else(|_| Value::Object(Default::default()))
}

fn write_cache_value(value: &Value) -> Result<(), StorageError> {
    let pruned = prune_course_pages(value.clone());
    write_json_pretty(cache_file()?, &pruned)
}

fn prune_course_pages(mut value: Value) -> Value {
    let pages = value.get("coursePages").cloned();
    if let Some(Value::Object(map)) = pages {
        let mut entries: Vec<(String, Value)> = map
            .into_iter()
            .filter(|(_, entry)| {
                entry
                    .get("timestamp")
                    .and_then(Value::as_u64)
                    .map(|ts| !is_expired(ts))
                    .unwrap_or(false)
            })
            .collect();
        entries.sort_by(|left, right| {
            right
                .1
                .get("timestamp")
                .and_then(Value::as_u64)
                .unwrap_or(0)
                .cmp(&left.1.get("timestamp").and_then(Value::as_u64).unwrap_or(0))
        });
        entries.truncate(MAX_CACHED_COURSE_PAGES);
        let map: serde_json::Map<String, Value> = entries.into_iter().collect();
        value["coursePages"] = Value::Object(map);
    }
    value
}

fn parse_courses(value: &Value) -> Vec<Course> {
    value
        .as_array()
        .map(|items| {
            items
                .iter()
                .filter_map(|entry| serde_json::from_value::<Course>(entry.clone()).ok())
                .collect()
        })
        .unwrap_or_default()
}

fn parse_upcoming(value: &Value) -> Vec<UpcomingAssignment> {
    value
        .as_array()
        .map(|items| {
            items
                .iter()
                .filter_map(|entry| serde_json::from_value::<UpcomingAssignment>(entry.clone()).ok())
                .collect()
        })
        .unwrap_or_default()
}

fn parse_sections(value: &Value) -> Vec<CourseSection> {
    value
        .as_array()
        .map(|items| {
            items
                .iter()
                .filter_map(|entry| serde_json::from_value::<CourseSection>(entry.clone()).ok())
                .collect()
        })
        .unwrap_or_default()
}

pub fn get_cached_dashboard() -> Option<DashboardCacheData> {
    let cache = read_cache_value();
    if let Some(dashboard) = cache.get("dashboard") {
        let timestamp = dashboard.get("timestamp").and_then(Value::as_u64)?;
        if is_expired(timestamp) {
            return None;
        }
        let data = dashboard.get("data")?;
        return Some(DashboardCacheData {
            courses: parse_courses(data.get("courses").unwrap_or(&Value::Null)),
            upcoming_assignments: parse_upcoming(
                data.get("upcomingAssignments").unwrap_or(&Value::Null),
            ),
        });
    }

    let legacy_timestamp = cache.get("timestamp").and_then(Value::as_u64);
    if let Some(ts) = legacy_timestamp {
        if is_expired(ts) {
            return None;
        }
    }
    let courses = parse_courses(cache.get("courses").unwrap_or(&Value::Null));
    if courses.is_empty() && !cache.get("courses").map(Value::is_array).unwrap_or(false) {
        return None;
    }
    Some(DashboardCacheData {
        courses,
        upcoming_assignments: Vec::new(),
    })
}

pub fn save_dashboard_to_cache(
    courses: &[Course],
    upcoming: &[UpcomingAssignment],
) -> Result<(), StorageError> {
    let mut cache = read_cache_value();
    let now = now_ms();
    cache["timestamp"] = json!(now);
    cache["courses"] = serde_json::to_value(courses)?;
    cache["dashboard"] = json!({
        "timestamp": now,
        "data": {
            "courses": courses,
            "upcomingAssignments": upcoming,
        }
    });
    write_cache_value(&cache)
}

pub fn save_courses_to_cache(courses: &[Course]) -> Result<(), StorageError> {
    let cached = get_cached_dashboard();
    let upcoming = cached.map(|d| d.upcoming_assignments).unwrap_or_default();
    save_dashboard_to_cache(courses, &upcoming)
}

pub fn get_cached_course_sections(course_id: i64) -> Option<Vec<CourseSection>> {
    let mut cache = read_cache_value();
    let key = course_id.to_string();
    let entry = cache.get("coursePages")?.get(&key)?.clone();
    let timestamp = entry.get("timestamp").and_then(Value::as_u64)?;
    if is_expired(timestamp) {
        if let Some(Value::Object(pages)) = cache.get_mut("coursePages") {
            pages.remove(&key);
        }
        let _ = write_cache_value(&cache);
        return None;
    }
    let data = entry.get("data")?;
    let sections = parse_sections(data);
    if sections.is_empty() && !data.is_array() {
        return None;
    }
    Some(sections)
}

pub fn save_course_sections_to_cache(
    course_id: i64,
    sections: &[CourseSection],
) -> Result<(), StorageError> {
    let mut cache = read_cache_value();
    let key = course_id.to_string();
    let pages = cache.get_mut("coursePages");
    let pages_obj = match pages {
        Some(Value::Object(map)) => map,
        _ => {
            cache["coursePages"] = Value::Object(Default::default());
            cache.get_mut("coursePages").unwrap().as_object_mut().unwrap()
        }
    };
    pages_obj.insert(
        key,
        json!({
            "timestamp": now_ms(),
            "data": sections,
        }),
    );
    write_cache_value(&cache)
}

pub fn clear_cache() -> Result<(), StorageError> {
    let now = now_ms();
    let payload = json!({
        "timestamp": now,
        "courses": [],
        "dashboard": {
            "timestamp": now,
            "data": { "courses": [], "upcomingAssignments": [] }
        },
        "coursePages": {},
    });
    write_json_pretty(cache_file()?, &payload)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Mutex;

    static ENV_LOCK: Mutex<()> = Mutex::new(());

    fn with_temp_dir<F: FnOnce()>(test: F) {
        let _guard = ENV_LOCK.lock().unwrap();
        let tempdir = tempfile::tempdir().unwrap();
        // SAFETY: tests serialised by ENV_LOCK
        unsafe { std::env::set_var(crate::storage::CONFIG_DIR_ENV, tempdir.path()); }
        test();
        unsafe { std::env::remove_var(crate::storage::CONFIG_DIR_ENV); }
    }

    #[test]
    fn round_trip_dashboard() {
        with_temp_dir(|| {
            let courses = vec![Course {
                id: 1,
                shortname: "MAT".into(),
                fullname: "Mathematics".into(),
                displayname: None,
                categoryid: None,
                categoryname: None,
                summary: None,
                visible: Some(1),
                progress: None,
                courseurl: None,
            }];
            let upcoming = vec![UpcomingAssignment {
                id: 9,
                name: "Hw1".into(),
                due_date: 1_700_000_000,
                course_id: 1,
                course_short_name: Some("MAT".into()),
                course_full_name: Some("Mathematics".into()),
            }];
            save_dashboard_to_cache(&courses, &upcoming).unwrap();
            let loaded = get_cached_dashboard().unwrap();
            assert_eq!(loaded.courses, courses);
            assert_eq!(loaded.upcoming_assignments, upcoming);
        });
    }

    #[test]
    fn course_pages_round_trip() {
        with_temp_dir(|| {
            let sections = vec![CourseSection {
                id: 10,
                name: Some("Week 1".into()),
                section: Some(1),
                summary: None,
                visible: Some(1),
                modules: vec![],
            }];
            save_course_sections_to_cache(42, &sections).unwrap();
            let loaded = get_cached_course_sections(42).unwrap();
            assert_eq!(loaded, sections);
        });
    }
}
